compiler/rustc_typeck/src/impl_wf_check/min_specialization.rs

```diff
@@ -368,25 +364,52 @@ fn check_predicates<'tcx>(
     );
 
     for (predicate, span) in impl1_predicates {
-        if !impl2_predicates.contains(&predicate) {
+        if !impl2_predicates.iter().any(|pred2| trait_predicates_eq(predicate, *pred2)) {
             check_specialization_on(tcx, predicate, span)
         }
     }
 }
 
+/// Checks whether two predicates are the same for the purposes of specialization.
+///
+/// This is slightly more complicated than simple syntactic equivalence, since
+/// we want to equate `T: Tr` with `T: ~const Tr` so this can work:
+///
+/// #[rustc_specialization_trait]
+/// trait Specialize { }
+///
+/// impl<T: ~const Bound> const Tr for T { }
+/// impl<T: Bound + Specialize> Tr for T { }
```

Comment:
```suggestion
/// ```rust
/// #[rustc_specialization_trait]
/// trait Specialize { }
///
/// impl<T: ~const Bound> const Tr for T { }
/// impl<T: Bound + Specialize> Tr for T { }
/// ```
```

Reply:
a bit confused by this, we have 1 const impl and specialize that using a non const impl :thinking: isn't that an issue because either:

- we use the non-const specialization during ctfe, which is unsound
- we don't use the non-const specialization during ctfe, which is at least unexpected

I would expect specializations to be at least as const as the impl as they are specializing

Reply:
What is ctfe?

Reply:
ctfe is the compile time function evaluator